content/knowledge-base/guides/write-a-composition-function-in-go.md

```diff
@@ -0,0 +1,825 @@
+---
+title: Write a Composition Function in Go
+state: beta
+alphaVersion: "1.11"
+betaVersion: "1.14"
+weight: 80
+description: "Composition functions allow you to template resources using Go"
+---
+
+Composition functions (or just functions, for short) are custom programs that
+template Crossplane resources. Crossplane calls composition functions to
+determine what resources it should create when you create a composite resource
+(XR). Read the
+[concepts](/latest/concepts/composition-functions)
+page to learn more about composition functions.
+
+You can write a function to template resources using a general purpose
+programming language. Using a general purpose programming language allows a
+Function to use more advanced logic to template resources, like loops and
+conditionals. This guide explains how to write a composition function in
+[Go](https://go.dev).
+
+{{< hint "important" >}}
+It helps to be familiar with
+[how composition functions work](/latest/concepts/composition-functions#how-composition-functions-work)
+before following this guide.
+{{< /hint >}}
+
+## Understand the steps
+
+This guide covers writing a composition function for an
+{{<hover label="xr" line="2">}}XBuckets{{</hover>}} composite resource (XR).
+
+```yaml {label="xr"}
+apiVersion: example.crossplane.io/v1
+kind: XBuckets
+metadata:
+  name: example-buckets
+spec:
+  region: us-east-2
+  names:
+  - crossplane-functions-example-a
+  - crossplane-functions-example-b
+  - crossplane-functions-example-c
+```
+
+<!-- vale gitlab.FutureTense = NO -->
+<!--
+This section is setting the stage for future sections. It doesn't make sense to
+refer to the function in the present tense, because it doesn't exist yet.
+-->
+An `XBuckets` XR has a region and an array of bucket names. The function will
+create an Amazon Web Services (AWS) S3 bucket for each entry in the names array.
+<!-- vale gitlab.FutureTense = NO -->
+
+To write a function in Go you:
+
+1. Install the tools you need to write the function
+1. Initialize the function from a template
+1. Edit the template to add the function's logic
+1. Test the function end-to-end
+1. Build and push the function to a package repository
+
+This guide covers each of these steps in detail.
+
+## Install the tools you need to write the function
+
+To write a function in Go you need:
+
+* [Go](https://go.dev/dl/) 1.21 or newer. The guide uses Go 1.21.
+* [Docker Engine](https://docs.docker.com/engine/). This guide uses Engine v24.
+* The [Crossplane CLI](https://github.com/crossplane/docs/pull/584) v1.14 or
```

Comment:
we can point to the actual merged CLI docs now, this points to a PR

Reply:
We can't. they're in the v1.14-release branch but not in master until https://github.com/crossplane/docs/pull/596 is merged.

Reply:
The CLI docs appeared, so they're now linked. :)